<?xml version="1.0"?>
<robot name="wp_v4">

  <!-- material -->
  <material name = "black">
    <color rgba = "0.01 0.01 0.01 1"/>
  </material>
  <material name = "grey">
    <color rgba = "0.2 0.2 0.2 1"/>
  </material>
  <material name = "blue">
    <color rgba = "0.2 0.2 0.9 1"/>
  </material>


<link name="base_footprint">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.05 0.05 0.001" />
      </geometry>
      <material name="TransparentGreen" />
    </visual>
    <collision>
      <origin xyz="0 0 .1" rpy="0 0 0" />
      <geometry>
        <box size="0.76 0.5 .2"/>
      </geometry>
    </collision>
    <inertial>
    <mass value="10"/>
    <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
</link>

<joint name="base_joint" type="fixed">
  <origin xyz="0 0 0.09" rpy="0 0 0" />
  <parent link="base_footprint"/>
  <child link="base_link" />
</joint>

  <!-- base -->
  <link name="base_link">
  <visual>
   <geometry>
    <cylinder length="0.6" radius="0.0"/>
   </geometry>
   <origin rpy = "0 0 0" xyz = "0 0 0"/>
  </visual>
  </link>

  <!-- body -->
  <link name = "body_link">
    <visual>
      <geometry>
        <mesh filename="package://wpv4/meshes/body.dae" scale="1 1 1"/>
      </geometry>
      <origin rpy = "1.57 0 1.57" xyz = "-0.38 -0.25 -0.02"/>
    </visual>
    <collision>
      <origin xyz="0 0 .15" rpy="0 0 0" />
      <geometry>
        <box size="0.76 0.5 0.095"/>
      </geometry>
    </collision>
    <inertial>
    <mass value="10"/>
    <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <joint name = "base_to_body" type = "fixed">
    <parent link = "base_link"/>
    <child link = "body_link"/>
    <origin rpy="0 0 0" xyz="0 0 0.02"/> <!--pos-->
  </joint>
  <gazebo reference = "body_link">
            <turnGravityOff>false</turnGravityOff>
  </gazebo>

  <!-- tread -->
  <link name = "tread_left">
    <visual>
      <geometry>
        <mesh filename="package://wpv4/meshes/tread_left.dae" scale="1 1 1"/>
      </geometry>
      <origin rpy = "1.57 0 0" xyz = "-0.55 0.085 -0.14"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="1 0.2 0.3 "/>
      </geometry>
    </collision>
    <inertial>
    <mass value="10"/>
    <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <joint name = "base_to_left" type = "fixed">
    <parent link = "base_link"/>
    <child link = "tread_left"/>
    <origin rpy="0 0 0" xyz="0.135 0.26 0.038"/>
 </joint>

  <link name = "tread_right">
  <visual>
    <geometry>
      <mesh filename="package://wpv4/meshes/tread_right.dae" scale="1 1 1"/>
    </geometry>
    <origin rpy = "1.57 0 3.14" xyz = "0.55 -0.05 -0.14"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="1.57 0 0" />
    <geometry>
      <box size="1 0.2 0.3 "/>
    </geometry>
  </collision>
  <inertial>
  <mass value="10"/>
  <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
  </inertial>
</link>
<joint name = "base_to_right" type = "fixed">
  <parent link = "base_link"/>
  <child link = "tread_right"/>
  <origin rpy="0 0 3.14" xyz="0.135 -0.26 0.038"/> 
</joint>
   

<!-- Lidar -->
<link name = "laser">
  <visual>
   <geometry>
    <cylinder length="0.00" radius="0.00"/>
   </geometry>
   <origin rpy = "0 0 0" xyz = "0 0 0"/>
  </visual>
</link>
<joint name="laser_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.2 0 0"  /> <!--pos-->
    <parent link="base_link" />
    <child link="laser" />
</joint>

<!-- velodyne -->
<link name = "velo_link">
  <visual>
   <geometry>
    <cylinder length="0.03" radius="0.03"/>
   </geometry>
   <origin rpy = "0 0 0" xyz = "0 0 0"/>
  </visual>
</link>
<joint name="velodyne_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.2 0 1.7"  /> <!--pos-->
    <parent link="base_link" />
    <child link="velo_link" />
</joint>

<!-- Kinect -->
<link name = "kinect2_ir_optical_frame">
  <visual>
   <geometry>
    <box size="0 0 0"/>
   </geometry>
   <origin rpy = "0 0 0" xyz = "0 0 0"/>
   <material name = "blue"/>
  </visual>
</link>
<joint name="kinect_joint" type="fixed">
    <origin xyz="0.2 0 0.2" rpy="-1.5707963 0 -1.5707963" /> <!--pos-->
    <parent link="base_link" />
    <child link="kinect2_ir_optical_frame" />
</joint>

<!-- Gazebo plugin for ROS Control -->
<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
    <robotNamespace>/</robotNamespace>
  </plugin>
</gazebo>

</robot>
